crates/checkpoint-downloader/src/downloader.rs

```diff
@@ -197,6 +197,49 @@ impl ParallelCheckpointDownloaderInner {
         Ok(())
     }
 
+    async fn adjust_workers(
+        next_worker_id: &mut usize,
+        current_workers: usize,
+        target_workers: usize,
+        channels: &PoolMonitorChannels,
+        config: &PoolMonitorConfig,
+        worker_count: &Arc<RwLock<usize>>,
+    ) -> Result<()> {
+        match current_workers.cmp(&target_workers) {
+            std::cmp::Ordering::Greater => {
+                let to_remove = current_workers - target_workers;
+                for _ in 0..to_remove {
+                    channels
+                        .message_sender
+                        .send(WorkerMessage::Shutdown)
+                        .await?;
```

Comment:
question: what if in the middle of removal, there is a failure, and returned here? Seems that worker_count won't be updated in this case.

Reply:
sure, fixed